builtin/v9/migration/miner.go

```diff
@@ -0,0 +1,141 @@
+package migration
+
+import (
+	"context"
+
+	"github.com/filecoin-project/go-state-types/builtin/v8/market"
+
+	"golang.org/x/xerrors"
+
+	commp "github.com/filecoin-project/go-commp-utils/nonffi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	miner8 "github.com/filecoin-project/go-state-types/builtin/v8/miner"
+	"github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
+	miner9 "github.com/filecoin-project/go-state-types/builtin/v9/miner"
+
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+
+	"github.com/filecoin-project/go-state-types/abi"
+)
+
+type minerMigrator struct {
+	proposals  *market.DealArray
+	OutCodeCID cid.Cid
+}
+
+func (m minerMigrator) migratedCodeCID() cid.Cid {
+	return m.OutCodeCID
+}
+
+func (m minerMigrator) migrateState(ctx context.Context, store cbor.IpldStore, in actorMigrationInput) (*actorMigrationResult, error) {
+	var inState miner8.State
+	if err := store.Get(ctx, in.head, &inState); err != nil {
+		return nil, err
+	}
+	wrappedStore := adt.WrapStore(ctx, store)
+
+	oldPrecommitOnChainInfos, err := adt.AsMap(wrappedStore, inState.PreCommittedSectors, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load old precommit onchain infos for miner %s: %w", in.address, err)
+	}
+
+	emptyMap, err := adt.StoreEmptyMap(wrappedStore, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to make empty map: %w", err)
+	}
+
+	newPrecommitOnChainInfos, err := adt.AsMap(wrappedStore, emptyMap, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load empty map: %w", err)
+	}
+
+	var info miner8.SectorPreCommitOnChainInfo
+	err = oldPrecommitOnChainInfos.ForEach(&info, func(key string) error {
+		var unsealedCid *cid.Cid
+		if len(info.Info.DealIDs) != 0 {
+			pieces := make([]abi.PieceInfo, len(info.Info.DealIDs))
+			for i, dealID := range info.Info.DealIDs {
+				deal, err := m.proposals.GetDealProposal(dealID)
+				if err != nil {
+					return xerrors.Errorf("error getting deal proposal: %w", err)
+				}
+
+				pieces[i] = abi.PieceInfo{
+					PieceCID: deal.PieceCID,
+					Size:     deal.PieceSize,
+				}
+			}
+
+			commd, err := generateUnsealedCID(info.Info.SealProof, pieces)
+			if err != nil {
+				return xerrors.Errorf("failed to generate unsealed CID: %w", err)
+			}
+
+			unsealedCid = &commd
+		}
+
+		err = newPrecommitOnChainInfos.Put(miner9.SectorKey(info.Info.SectorNumber), &miner9.SectorPreCommitOnChainInfo{
+			Info: miner9.SectorPreCommitInfo{
+				SealProof:     info.Info.SealProof,
+				SectorNumber:  info.Info.SectorNumber,
+				SealedCID:     info.Info.SealedCID,
+				SealRandEpoch: info.Info.SealRandEpoch,
+				DealIDs:       info.Info.DealIDs,
+				Expiration:    info.Info.Expiration,
+				UnsealedCid:   unsealedCid,
```

Comment:
All looks good. Idly wondering why is this is a `*cid.Cid` as opposed to `cid.Cid`. Is it to allow for the optional/assumed CC CommD?